# What outliers, strange, and missing values did you find on the trips table? What would you do to deal with this data and why?

In [1]:
#Importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Trips Table

In [2]:
#Importing trips table
trips_table = pd.read_csv("Datasets/trip.csv")
trips_table

tripid    personid trip_path_distance  speed_mph  \
0      1710000501001  1710000501        2.299694071  13.798164   
1      1710000501002  1710000501        1.122817397  13.473809   
2      1710000501003  1710000501        3.263440492  19.580643   
3      1710000501004  1710000501        8.126289938  19.503096   
4      1710000501005  1710000501        8.044890337  24.134671   
5      1710000502001  1710000502        2.299694071  13.798164   
6      1710000502002  1710000502        1.122817397  13.473809   
7      1710000502003  1710000502        3.263440492  19.580643   
8      1710002401001  1710002401        0.541214141   3.247285   
9      1710002401002  1710002401        0.541214141   3.247285   
10     1710002401003  1710002401        3.309421946  13.237688   
11     1710002401004  1710002401        3.309421946   9.928266   
12     1710002401005  1710002401        0.151614524   1.819374   
13     1710002401006  1710002401        0.151614524   1.819374   
14     1710002401007  1710002401        0.830773027   4.984638   
15     1710002401008  1710002401        0.923357306   5.540144   
16     1710002402001  1710002402        0.799083106   4.794499   
17     1710002402002  1710002402        0.891046014   5.346276   
18     1710002402003  1710002402        1.405541202   3.373299   
19     1710002402004  1710002402        1.546592419   4.639777   
20     1710002403001  1710002403        0.799083106   4.794499   
21     1710002403002  1710002403        0.923357306   5.540144   
22     1710005201001  1710005201        1.509310159   3.018620   
23     1710005201002  1710005201        0.505174623   1.515524   
24     1710005201003  1710005201        0.411347602   1.645390   
25     1710005201004  1710005201         0.71457665   1.714984   
26     1710005201005  1710005201        1.097341186   1.881156   
27     1710005901001  1710005901        8.689254752  12.124542   
28     1710005901002  1710005901        0.647468782   3.884813   
29     1710005901003  1710005901        0.841336594   5.608911   
...              ...         ...                ...        ...   
52462  1715410801004  1715410801        1.102933525   3.308801   
52463  1715412901001  1715412901        0.612671996   2.625737   
52464  1715412901002  1715412901        0.471620735   2.829724   
52465  1715412901003  1715412901        4.737955341  14.961964   
52466  1715412901004  1715412901        5.111399427  13.334085   
52467  1715412901005  1715412901        8.422686511  24.064819   
52468  1715412901006  1715412901        3.485271017  17.426355   
52469  1715412901007  1715412901         3.26282013  16.314101   
52470  1715412901008  1715412901        1.047010459   2.512825   
52471  1715412901009  1715412901        8.164817466  13.240245   
52472  1715412901010  1715412901        4.436590313  14.010285   
52473  1715412901011  1715412901        3.680381572  14.721526   
52474  1715412901012  1715412901        3.112448302  13.339064   
52475  1715412901013  1715412901        3.620729937  15.517414   
52476  1715412901014  1715412901          0.7506164   2.814812   
52477  1715412902001  1715412902        3.313772568  18.075123   
52478  1715412902002  1715412902        3.196333413  17.434546   
52479  1715412902003  1715412902        1.408027122   3.379265   
52480  1715412902004  1715412902          7.9634932  14.053223   
52481  1715412902005  1715412902        4.426648373  12.647567   
52482  1715412902006  1715412902        3.659254951  15.682521   
52483  1715412902007  1715412902        0.549292134   1.938678   
52484  1715412902008  1715412902        3.083243856  13.213902   
52485  1715412902009  1715412902        3.574748469  19.498628   
52486  1715412902010  1715412902        0.625099419   2.344123   
52487  1715419901001  1715419901        31.44758631  20.965058   
52488  1715419901002  1715419901         32.6219775  19.573186   
52489  1715419901003  1715419901                0.2   2.400000   
52490  1715419901004  1715419901        0.665488341   

### Finding Missing Values Based On Distance of the Trip Path

In [5]:
missing_values_trip = trips_table[trips_table["trip_path_distance"] == " "]
missing_values_trip

tripid    personid trip_path_distance  speed_mph  \
14892  1711494102001  1711494102                           NaN   
16459  1711625701001  1711625701                           NaN   
22942  1712399302001  1712399302                           NaN   
23793  1712469601001  1712469601                           NaN   
40677  1714124802001  1714124802                           NaN   
45650  1714618802001  1714618802                           NaN   
48788  1714997602001  1714997602                           NaN   

                                     Destination purpose  \
14892  Went to work-related place (e.g., meeting, sec...   
16459                                          Went home   
22942                                          Went home   
23793  Attended recreational event (e.g., movies, spo...   
40677  Went to school/daycare (e.g., daycare, K-12, c...   
45650                          Went to primary workplace   
48788                          Went to primary workplace   

                                            Primary Mode  
14892                             Airplane or helicopter  
16459                             Airplane or helicopter  
22942                             Airplane or helicopter  
23793                             Airplane or helicopter  
40677  Other mode (e.g., skateboard, kayak, motorhome...  
45650       Urban rail (e.g., Link light rail, monorail)  
48788                            Other household vehicle

Much of the missing values for trip path distance usually had a primary mode of transport related to flight, such as an airplane or a helicopter. The lack of distance given would often make the speed of the trip unknown as well.

## Missing Values

### Finding Missing Values Based on Speed

In [19]:
missing_values_speed = pd.isnull(trips_table["speed_mph"])
ms_speed = trips_table[missing_values_speed]
ms_speed

tripid    personid trip_path_distance  speed_mph  \
7620   1710768901023  1710768901        0.564826414        NaN   
14892  1711494102001  1711494102                           NaN   
16459  1711625701001  1711625701                           NaN   
19748  1712029601006  1712029601        1.856657122        NaN   
22373  1712340301016  1712340301        0.318763422        NaN   
22942  1712399302001  1712399302                           NaN   
23793  1712469601001  1712469601                           NaN   
28335  1712919601011  1712919601        0.707120417        NaN   
28336  1712919601012  1712919601        0.784791816        NaN   
28463  1712930501010  1712930501         17.7022439        NaN   
32939  1713292401028  1713292401        0.821452716        NaN   
32948  1713292401037  1713292401        0.825180943        NaN   
33005  1713292403013  1713292403        0.821452716        NaN   
33011  1713292403019  1713292403        0.825180943        NaN   
40677  1714124802001  1714124802                           NaN   
42653  1714320702027  1714320702        2.761373578        NaN   
45067  1714573001047  1714573001        0.285830748        NaN   
45650  1714618802001  1714618802                           NaN   
48788  1714997602001  1714997602                           NaN   

                                     Destination purpose  \
7620      Went to other shopping (e.g., mall, pet store)   
14892  Went to work-related place (e.g., meeting, sec...   
16459                                          Went home   
19748                                                NaN   
22373                              Went grocery shopping   
22942                                          Went home   
23793  Attended recreational event (e.g., movies, spo...   
28335                                                NaN   
28336                                                NaN   
28463  Dropped off/picked up someone (e.g., son at a ...   
32939  Dropped off/picked up someone (e.g., son at a ...   
32948  Dropped off/picked up someone (e.g., son at a ...   
33005  Dropped off/picked up someone (e.g., son at a ...   
33011  Dropped off/picked up someone (e.g., son at a ...   
40677  Went to school/daycare (e.g., daycare, K-12, c...   
42653                                                NaN   
45067  Dropped off/picked up someone (e.g., son at a ...   
45650                          Went to primary workplace   
48788                          Went to primary workplace   

                                            Primary Mode  
7620                                 Household vehicle 1  
14892                             Airplane or helicopter  
16459                             Airplane or helicopter  
19748                                                NaN  
22373                           Walk, jog, or wheelchair  
22942                             Airplane or helicopter  
23793                             Airplane or helicopter  
28335                                                NaN  
28336                                                NaN  
28463                                Household vehicle 1  
32939                                Household vehicle 1  
32948                                Household vehicle 1  
33005                                Household vehicle 1  
33011                                Household vehicle 1  
40677  Other mode (e.g., skateboard, kayak, motorhome...  
42653                                                NaN  
45067                             Private bus or shuttle  
45650       Urban rail (e.g., Link light rail, monorail)  
48788                            Other household vehicle

For the NaN values for speed that had distance present, most trips had a distance of under a mile, with one trip having an outlier of 17 miles. The mode of the purpose of the trips were from dropping off or picking up someone.

### Finding Missing Values Based on Destination Purpose

In [21]:
missing_values_destination = pd.isnull(trips_table["Destination purpose"])
trips_table[missing_values_destination]

tripid    personid trip_path_distance     speed_mph  \
996    1710083402033  1710083402         4.78455818     31.897055   
997    1710083402034  1710083402        4.078680506     20.393403   
998    1710083402035  1710083402        0.489640499      4.896405   
1771   1710137502006  1710137502        0.171498449      3.429969   
1799   1710137502034  1710137502        0.175226676      2.628400   
2193   1710187302011  1710187302        2627.303423    489.559644   
2200   1710187302018  1710187302        6.747469777     36.804381   
2201   1710187302019  1710187302         20.6096397    112.416217   
2203   1710187302021  1710187302        251.5695836    457.399243   
2204   1710187302022  1710187302        807.6707031    598.274595   
2205   1710187302023  1710187302        599.2429213  11984.858430   
2206   1710187302024  1710187302        535.5256552    868.419981   
2209   1710187302027  1710187302        0.935785016      9.357850   
2558   1710222701028  1710222701        2.255577428     15.037183   
2621   1710226902001  1710226902        0.798461982     15.969240   
2622   1710226902002  1710226902          0.5629623     11.259246   
2643   1710226902023  1710226902        0.497718325      9.954366   
2644   1710226902024  1710226902         1.17252744     23.450549   
2646   1710226902026  1710226902        0.494611469      7.419172   
2650   1710226902030  1710226902        3.533116599     30.283857   
2653   1710226902033  1710226902        1.609972759      6.899883   
2654   1710226902034  1710226902        28.46625706     36.339903   
2656   1710226902036  1710226902        0.454843713      2.729062   
2721   1710227501049  1710227501        0.188896842      2.833453   
2733   1710227502011  1710227502        1.085535473     13.026426   
2734   1710227502012  1710227502        1.092991927      7.286613   
3005   1710254902014  1710254902        0.370337231      5.555058   
3138   1710261801026  1710261801        0.126759723      2.535194   
3139   1710261801027  1710261801         0.28645212      4.296782   
3155   1710262801014  1710262801         20.6096397     41.219279   
...              ...         ...                ...           ...   
43331  1714373302010  1714373302        13.88018969     34.700474   
43332  1714373302011  1714373302        1.262004891     10.817185   
43333  1714373302012  1714373302        2.218916527     14.792777   
43334  1714373302013  1714373302        5.723450052     26.415923   
43335  1714373302014  1714373302        30.66839656     41.820541   
43336  1714373302015  1714373302        5.929745288     20.928513   
43337  1714373302016  1714373302        3.014271654     16.441482   
43897  1714429003007  1714429003        0.445523145      3.818770   
43904  1714429003014  1714429003         2.05860276      7.265657   
43905  1714429003015  1714429003        8.454997813     12.682497   
43906  1714429003016  1714429003        0.812132148      9.745586   
43907  1714429003017  1714429003        1.588224767      7.941124   
43908  1714429003018  1714429003        1.892696652     11.356180   
43909  1714429003019  1714429003        3.888540921      3.763104   
43910  1714429003020  1714429003        6.080738487     24.322954   
45845  1714649402018  1714649402        79.73310865   1594.662173   
46962  1714787902006  1714787902        98.27979599     60.171304   
46963  1714787902007  1714787902        7.818092341     20.395023   
47942  1714928501003  1714928501        0.585953034     11.719061   
47991  1714928502017  1714928502         0.76117971     11.417696   
48000  1714928502026  1714928502        0.935785016      8.021014   
48001  1714928502027  1714928502        1.810054283      7.757375   
48701  1714994101014  1714994101        0.444280402      2.961869   
48704  1714994101017  1714994101        124.6824793     17.118876   
48722  1714994102018  1714994102        0.207537978      4.150760   
48723  1714994102019  1714994102        0.147886344      2.957727   
48724  17149

The vast majority of the missing values for the trip purpose also seemed to have missing values for the primary mode. An outlier in the case of a missing trip purpose with be a present value reported for primary mode of getting around.

### Finding Missing Values Based on Primary Mode

In [28]:
missing_values_primary = pd.isnull(trips_table["Primary Mode"])
trips_table[missing_values_primary]

tripid    personid trip_path_distance    speed_mph  \
565    1710059001031  1710059001        0.778578104     9.342937   
566    1710059001032  1710059001        0.840093852    10.081126   
567    1710059001033  1710059001        2.614108606    15.684652   
568    1710059001034  1710059001        2.515310586    16.768737   
569    1710059001035  1710059001        2.950891792    17.705351   
570    1710059001036  1710059001        2.691158634    12.420732   
571    1710059001037  1710059001        2.696129603     6.470711   
572    1710059001038  1710059001                0.4     2.666667   
573    1710059001039  1710059001                0.4     3.000000   
574    1710059001040  1710059001        2.070408813     4.968981   
575    1710062701001  1710062701        2.064815833     8.259263   
576    1710062701002  1710062701        2.171691645     8.686767   
577    1710062701003  1710062701         1.48507669     4.455230   
578    1710062701004  1710062701        1.486319432     3.567167   
579    1710062701005  1710062701        0.318141952     1.908852   
580    1710062701006  1710062701        0.319384694     1.916308   
581    1710062701007  1710062701        6.230487017    37.382922   
582    1710062701008  1710062701        4.407384503    17.629538   
583    1710065201001  1710065201        0.690964766     8.291577   
584    1710065201002  1710065201        0.543078422    10.861568   
585    1710065201003  1710065201         6.87982184    29.484951   
586    1710065201004  1710065201        6.684711286    40.108268   
587    1710065201005  1710065201        0.942620099     8.079601   
588    1710065201006  1710065201        2.456280323    13.397893   
589    1710065201007  1710065201        1.065651595    15.984774   
590    1710065201008  1710065201        0.966853575    14.502804   
591    1710065201009  1710065201        0.459193311     4.591933   
592    1710065201010  1710065201        1.555913465     9.335481   
593    1710065201011  1710065201        3.713935616    13.927259   
594    1710065201012  1710065201        3.261577388    16.307887   
...              ...         ...                ...          ...   
43333  1714373302012  1714373302        2.218916527    14.792777   
43334  1714373302013  1714373302        5.723450052    26.415923   
43335  1714373302014  1714373302        30.66839656    41.820541   
43336  1714373302015  1714373302        5.929745288    20.928513   
43337  1714373302016  1714373302        3.014271654    16.441482   
43338  1714373302017  1714373302        33.85043844    48.357769   
43897  1714429003007  1714429003        0.445523145     3.818770   
43904  1714429003014  1714429003         2.05860276     7.265657   
43905  1714429003015  1714429003        8.454997813    12.682497   
43906  1714429003016  1714429003        0.812132148     9.745586   
43907  1714429003017  1714429003        1.588224767     7.941124   
43908  1714429003018  1714429003        1.892696652    11.356180   
43909  1714429003019  1714429003        3.888540921     3.763104   
43910  1714429003020  1714429003        6.080738487    24.322954   
43911  1714429003021  1714429003        11.51960053    31.417092   
45845  1714649402018  1714649402        79.73310865  1594.662173   
46962  1714787902006  1714787902        98.27979599    60.171304   
46963  1714787902007  1714787902        7.818092341    20.395023   
47942  1714928501003  1714928501        0.585953034    11.719061   
47991  1714928502017  1714928502         0.76117971    11.417696   
48000  1714928502026  1714928502        0.935785016     8.021014   
48001  1714928502027  1714928502        1.810054283     7.757375   
48701  1714994101014  1714994101        0.444280402     2.961869   
48704  1714994101017  1714994101        124.6824793    17.118876   
48722  1714994102018  1714994102        0.207537978     4.150760   
48723  1714994102019  1714994102        0.147886344     2.957727   
48724  1714994102020  1714994102        0.169634335     3.392687   
48

Much of the missing data for primary home also had missing data for destination purpose, though much of the destination purpose answers that didnt have missing data reported the point of their trips was to return home. 

If both the destination purpose as well as the Primary Mode of transportation is missing, I would remove that entire row from the dataset as it lacks some crucial data that was needed for the survey. If I wanted to know the primary mode of transporation for each trip, any trip with a missing value would be a hindrance when comparing modes of transportation.

## Finding Outliers

### Finding Outliers Based on Distance of Trip

In [53]:
#Find the average distance of all trips
path_distance = trips_table["trip_path_distance"]

#Replacing empty strings with NaN values
path_distance.replace(' ', np.nan, inplace =True)
path_distance.dropna( inplace = True)

#Convert string to numeric
numeric = pd.to_numeric(path_distance)

#Finding the average
avg = numeric.mean()
avg

9.679149494596356

In [65]:
#Checking for distances over 20 miles
over_20 = numeric[numeric > 20]
over_20.count()

2521

In [66]:
over_100 = numeric[numeric > 100]
over_100.count()

272

In [73]:
over_1000 = numeric[numeric > 1000]
over_1000.count()

83

In [74]:
over_3000 = numeric[numeric > 3000]
over_3000.count()


10

There are 272 trips that covered over 100 miles, and 83 that covered over a thousand miles. Since the average distance of all trips taken is around 10 miles, taking into account other priamry modes of transportation, trips that exceeded 100 miles seem to be outliers and only account for 0.5% of all trips. Strangely enough, some trips reported over 3000 miles covered, the longest distance being over 6000 miles in a single trip.

### Finding Outliers Based on Speed of Trip

In [75]:
#Find the average speed of all trips
path_speed = trips_table["speed_mph"]

#Replacing empty strings with NaN values
path_speed.replace(' ', np.nan, inplace =True)
path_speed.dropna( inplace = True)

#Convert string to numeric
numeric_speed = pd.to_numeric(path_speed)

#Finding the average
avg_speed = numeric_speed.mean()
avg_speed

17.21969107087986

In [77]:
#Checking for speeds under 1
under_1 = numeric_speed[numeric_speed < 1]
under_1.count()

703

In [79]:
#Checking for speeds over 50
over_50 = numeric_speed[numeric_speed > 50]
over_50.count()

674

### Checking Outliers in Purpose of Trip

In [85]:
destination_count = trips_table["Destination purpose"].value_counts()
destination_count

Went home                                                                            16857
Went to primary workplace                                                             6604
Went to restaurant to eat/get take-out                                                4336
Dropped off/picked up someone (e.g., son at a friend's house, spouse at bus stop)     3173
Went grocery shopping                                                                 3095
Went to exercise (e.g., gym, walk, jog, bike ride)                                    2945
Went to other shopping (e.g., mall, pet store)                                        2546
Went to work-related place (e.g., meeting, second job, delivery)                      2073
Attended social event (e.g., visit with friends, family, co-workers)                  1983
Conducted personal business (e.g., bank, post office)                                 1811
Attended recreational event (e.g., movies, sporting event)                            1252

In [89]:
destination_count_under_300 = destination_count[destination_count < 300]
destination_count_under_300

Other social/leisure (rMove only)      246
Went to other work-related activity    138
Went to the zoo                          1
Name: Destination purpose, dtype: int64

### Checking for Outliers in Primary Mode of Transportation

In [87]:
primary_mode_count = trips_table["Primary Mode"].value_counts()
primary_mode_count

Household vehicle 1                                          22328
Walk, jog, or wheelchair                                     10827
Household vehicle 2                                           6787
Bus (public transit)                                          3843
Friend/colleague's car                                        1261
Bicycle or e-bike                                             1199
Other household vehicle                                        724
Household vehicle 3                                            715
Urban rail (e.g., Link light rail, monorail)                   646
Car from work                                                  593
Other hired service (e.g., Lyft, Uber)                         563
Rental car                                                     370
School bus                                                     354
Other non-household vehicle                                    242
Private bus or shuttle                                        

In [90]:
primary_mode_count_under_100 = primary_mode_count[primary_mode_count < 100]
primary_mode_count_under_100

Household vehicle 4                93
Taxi (e.g., Yello Cab)             60
Commuter rail (Sounder, Amtrak)    52
Other rail (e.g., streetcar)       28
Paratransit                        17
Other motorcycle/moped/scooter     13
Household vehicle 6                 2
Hopping                             1
Skateboard                          1
Name: Primary Mode, dtype: int64